### PR TITLE
Passive Checks

### PR DESCRIPTION
**Pull request by badrAZ**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,11 +12,15 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-nagios):
 
 ## Usage
 
-**TODO**
+Like all other xo-server plugins, it can be configured directly via
+the web iterface, see [the plugin documentation](https://xen-orchestra.com/docs/plugins.html). You can also test the configuration plugin if it works.
```
Comment (pdonias) [inline README.md]: interface*

Comment (pdonias) [reply README.md]: "You can also test if the configuration plugin works."

File: src/index.js [hunk 2]
```diff
@@ -1,26 +1,135 @@
...
+  test (data) {
+    return this._sendPassiveCheck(data.status, 'The server-nagios plugin for Xen Orchestra server seems to be working fine, nicely done :)')
+  }
+
+  _sendPassiveCheck (status, message) {
+    const client = new net.Socket()
+    client.connect(this._conf.port, this._conf.server, () => {
+      console.log('Successful connection')
+    })
+    client.on('data', (data) => {
```
Comment (pdonias) [inline src/index.js]: Parenthesis are not necessary when there is only one parameter.

File: src/index.js [hunk 3]
```diff
@@ -1,26 +1,135 @@
...
+      const xorPaquetBuffer = new Buffer(this._xor(this._xor(this._stringToAsciiArray(paquet.toString(encoding)), this._stringToAsciiArray(iv)), this._stringToAsciiArray(this._conf.key)), encoding)
+      client.write(xorPaquetBuffer, function (a) {
+        client.destroy()
+      })
+    })
+  }
+
+  _nscaPaquetBuilder (timestamp, iv, status, message, encoding) {
+    const size = 720
+    const version = 3
```
Comment (pdonias) [inline src/index.js]: `size` and `version` could be declared as global variables.

File: src/index.js [hunk 4]
```diff
@@ -1,26 +1,135 @@
...
+  _stringToAsciiArray (str) {
+    const map = Array.prototype.map
+    const chars = map.call(str, (x) => {
+      return x.charCodeAt(0)
+    })
+    return chars
+  }
+
+  _xor (long, short) {
+    let result = []
```
Comment (pdonias) [inline src/index.js]: Declare as `const`

File: src/index.js [hunk 5]
```diff
@@ -1,26 +1,135 @@
...
+    paquet.write(this._conf.host, 14, 77, encoding)
+    paquet.write(this._conf.service, 78, 206, encoding)
+    paquet.write(message, 206, size, encoding)
+    paquet.writeUInt32BE(crc32.unsigned(paquet), 4)
+    return paquet
+  }
+
+  _stringToAsciiArray (str) {
+    const map = Array.prototype.map
+    const chars = map.call(str, (x) => {
```
Comment (pdonias) [inline src/index.js]: Parenthesis.

File: README.md [hunk 6]
```diff
@@ -12,11 +12,15 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-nagios):
 
 ## Usage
 
-**TODO**
+Like all other xo-server plugins, it can be configured directly via
+the web interface, see [the plugin documentation](https://xen-orchestra.com/docs/plugins.html). You can also test if the configuration plugin works.
```
Comment (julien-f) [inline README.md]: Remove the second sentence, it's not important to write this here.

File: README.md [hunk 7]
```diff
@@ -12,11 +12,15 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-nagios):
...
 
 ## Development
 
-```
+### `Xo#sendPassiveCheck(status, message) `
+
+This xo method is called to send a passive check to nagios and change the status of a service.
```
Comment (julien-f) [inline README.md]: This does not explain what `status` and `message` should be.

File: src/index.js [hunk 8]
```diff
@@ -1,26 +1,135 @@
...
+    const client = new net.Socket()
+    client.connect(this._conf.port, this._conf.server, () => {
+      console.log('Successful connection')
+    })
+    client.on('data', data => {
+      const encoding = 'binary'
+      const dataBuffer = new Buffer(data)
+      const iv = dataBuffer.toString(encoding, 0, 128) // initialization vector
+      const timestamp = dataBuffer.readInt32BE(128)
+      const paquet = this._nscaPaquetBuilder(SIZE, VERSION, timestamp, iv, status, message, encoding)
```
Comment (olivierlambert) [inline src/index.js]: Use `packet` word and not `paquet`

File: src/index.js [hunk 9]
```diff
@@ -1,26 +1,135 @@
+import { Buffer } from 'buffer'
```
Comment (julien-f) [inline src/index.js]: Named imports after default imports. (use SublimeText sort algorithm as a reference)

File: src/index.js [hunk 10]
```diff
@@ -1,26 +1,135 @@
...
+import net from 'net'
+import crc32 from 'buffer-crc32'
```
Comment (julien-f) [inline src/index.js:1]: Keep them sorted.

File: src/index.js [hunk 11]
```diff
@@ -1,26 +1,135 @@
...
+    })
+    client.on('data', data => {
+      const encoding = 'binary'
+      const dataBuffer = new Buffer(data)
+      const iv = dataBuffer.toString(encoding, 0, 128) // initialization vector
+      const timestamp = dataBuffer.readInt32BE(128)
+      const paquet = this._nscaPaquetBuilder(SIZE, VERSION, timestamp, iv, status, message, encoding)
+      // 1) Using xor between the NSCA paquet and the initialization vector
+      // 2) Using xor between the result of the first operation and the encryption key
+      const xorPaquetBuffer = new Buffer(this._xor(this._xor(this._stringToAsciiArray(paquet.toString(encoding)), this._stringToAsciiArray(iv)), this._stringToAsciiArray(this._conf.key)), encoding)
```
Comment (olivierlambert) [inline src/index.js]: Same here, use `Packet`

File: src/index.js [hunk 12]
```diff
@@ -1,26 +1,135 @@
...
+      client.write(xorPaquetBuffer, function (a) {
+        client.destroy()
+      })
+    })
+  }
+
+  _nscaPaquetBuilder (size, version, timestamp, iv, status, message, encoding) {
```
Comment (olivierlambert) [inline src/index.js]: Change all `Paquet` occurrences by `Packet`

File: src/index.js [hunk 13]
```diff
@@ -1,26 +1,135 @@
...
+  type: 'object',
+
+  properties: {
+    server: {
+      type: 'string',
+      description: 'The nagios server adress'
+    },
+    port: {
+      type: 'integer',
+      description: 'The snca port'
```
Comment (julien-f) [inline src/index.js]: *snca* should be in uppercase, right?

File: src/index.js [hunk 14]
```diff
@@ -1,26 +1,135 @@
...
+    },
+    service: {
+      type: 'string',
+      description: 'The service description in Nagios'
+    }
+  },
+  additionalProperties: false,
+  required: ['server', 'port', 'key', 'host', 'service']
 }
+export const testSchema = {
```
Comment (julien-f) [inline src/index.js]: Add a blank line between these exports.

**Review by olivierlambert — CHANGES_REQUESTED** [review]

File: src/index.js [hunk 15]
```diff
@@ -1,26 +1,135 @@
...
+  type: 'object',
 
+  properties: {
+    status: {
```
Comment (julien-f) [inline src/index.js]: Is it necessary?

File: src/index.js [hunk 16]
```diff
@@ -1,26 +1,135 @@
...
+      type: 'integer',
+      description: 'The service status in Nagios (0: OK | 1: WARNING | 2: CRITICAL)'
+    }
+  },
+  additionalProperties: false,
+  required: ['status']
+}
+
+// ===================================================================
+const SIZE = 720
```
Comment (julien-f) [inline src/index.js]: Blank line before and *after* the separator.

File: src/index.js [hunk 17]
```diff
@@ -1,26 +1,135 @@
...
+const VERSION = 3
 class XoServerNagios {
```
Comment (julien-f) [inline src/index.js:86]: Blank line before the class.

File: src/index.js [hunk 18]
```diff
@@ -1,26 +1,135 @@
...
   unload () {
-    throw new Error('not implemented')
+    this._unset()
+  }
+
+  test ({status}) {
+    return this._sendPassiveCheck(status, 'The server-nagios plugin for Xen Orchestra server seems to be working fine, nicely done :)')
+  }
+
+  _sendPassiveCheck (status, message) {
```
Comment (julien-f) [inline src/index.js]: This method should returns a promise which resolves when the message has been sent or rejects in case of errors.

Comment (julien-f) [inline src/index.js]: Use named parameters for future compatibility.

File: src/index.js (hunk 11 above)
Comment (julien-f) [inline src/index.js]: Break into multiple lines and use indentation to help readability.

File: src/index.js (hunk 12 above)
Comment (julien-f) [inline src/index.js]: Should not be a method as it does not use `this`.

File: src/index.js [hunk 19]
```diff
@@ -1,26 +1,135 @@
...
+    paquet.writeUInt32BE(timestamp, 8)
+    paquet.writeInt16BE(status, 12)
+    paquet.write(this._conf.host, 14, 77, encoding)
+    paquet.write(this._conf.service, 78, 206, encoding)
+    paquet.write(message, 206, size, encoding)
+    paquet.writeUInt32BE(crc32.unsigned(paquet), 4)
+    return paquet
+  }
+
+  _stringToAsciiArray (str) {
```
Comment (julien-f) [inline src/index.js]: Not a method.

File: src/index.js [hunk 20]
```diff
@@ -1,26 +1,135 @@
...
+    const map = Array.prototype.map
+    const chars = map.call(str, x => {
+      return x.charCodeAt(0)
+    })
+    return chars
+  }
+
+  _xor (long, short) {
```
Comment (julien-f) [inline src/index.js]: Not a method.

File: src/index.js (hunk 19 above)
Comment (julien-f) [inline src/index.js]: Why not simply `string.split('')`?

**Review by julien-f — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 15 above)
Comment (badrAZ) [reply src/index.js]: I think it's not necessary, i will remove the testSchema and put a default status value.

File: src/index.js (hunk 13 above)
Comment (badrAZ) [reply src/index.js]: Yes, i will fix it.

File: src/index.js (hunk 19 above)
Comment (badrAZ) [reply src/index.js]: I wish convert the string to Ascii code

Comment (julien-f) [reply src/index.js]: `Buffer.from(string, 'ascii')`.

File: src/index.js [hunk 21]
```diff
@@ -1,26 +1,135 @@
...
+  }
+
+  _sendPassiveCheck (status, message) {
+    const client = new net.Socket()
+    client.connect(this._conf.port, this._conf.server, () => {
+      console.log('Successful connection')
+    })
+    client.on('data', data => {
+      const encoding = 'binary'
+      const dataBuffer = new Buffer(data)
```
Comment (julien-f) [inline src/index.js]: What is the initial type of `data`?
Why creating a new Buffer?

**Review by julien-f — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 21 above)
Comment (badrAZ) [reply src/index.js]: It's an error, the initial type of data is a buffer. I will fix it.

File: src/index.js [hunk 22]
```diff
@@ -32,22 +32,44 @@ export const configurationSchema = {
...
+  packet.writeInt16BE(version, 0)
+  packet.fill('h', 2, 3)
+  packet.writeUInt32BE(0, 4) // initial 0 for CRC32 value
+  packet.writeUInt32BE(timestamp, 8)
+  packet.writeInt16BE(status, 12)
+  packet.write(host, 14, 77, encoding)
+  packet.write(service, 78, 206, encoding)
+  packet.write(message, 206, size, encoding)
+  packet.writeUInt32BE(crc32.unsigned(packet), 4)
+  packet = Buffer.from(packet.toString(encoding), 'ascii')
```
Comment (julien-f) [inline src/index.js]: Why this line???

File: src/index.js [hunk 23]
```diff
@@ -32,22 +32,44 @@ export const configurationSchema = {
...
 
-  properties: {
-    status: {
-      type: 'integer',
-      description: 'The service status in Nagios (0: OK | 1: WARNING | 2: CRITICAL)'
+// ===================================================================
+
+function nscaPacketBuilder (host, service, size, version, timestamp, iv, status, message, encoding) {
+  // Building NSCA packet
+  let packet = new Buffer(size)
```
Comment (julien-f) [inline src/index.js]: `const`

Comment (julien-f) [inline src/index.js]: Wouldn't it be possible to automatically compute the size from the arguments?

File: src/index.js [hunk 24]
```diff
@@ -32,22 +32,44 @@ export const configurationSchema = {
...
-export const testSchema = {
-  type: 'object',
 
-  properties: {
-    status: {
-      type: 'integer',
-      description: 'The service status in Nagios (0: OK | 1: WARNING | 2: CRITICAL)'
+// ===================================================================
+
+function nscaPacketBuilder (host, service, size, version, timestamp, iv, status, message, encoding) {
```
Comment (julien-f) [inline src/index.js]: Use named parameters (an object), when there are so many of them.

File: src/index.js [hunk 25]
```diff
@@ -32,22 +32,44 @@ export const configurationSchema = {
...
+  packet.writeInt16BE(status, 12)
+  packet.write(host, 14, 77, encoding)
+  packet.write(service, 78, 206, encoding)
+  packet.write(message, 206, size, encoding)
+  packet.writeUInt32BE(crc32.unsigned(packet), 4)
+  packet = Buffer.from(packet.toString(encoding), 'ascii')
+  return packet
+}
+
+function xor (long, short) {
```
Comment (julien-f) [inline src/index.js]: I don't like the param names: the second one does not have to be shorter than the first one.

Use `data` and `mask` instead.

File: src/index.js [hunk 26]
```diff
@@ -32,22 +32,44 @@ export const configurationSchema = {
...
+  const result = []
```
Comment (julien-f) [inline src/index.js]: Why using an array instead of a buffer?

File: src/index.js [hunk 27]
```diff
@@ -32,22 +32,44 @@ export const configurationSchema = {
...
+  let j = 0
+  for (let i = 0; i < long.length; i++) {
+    if (j === short.length) {
```
Comment (julien-f) [inline src/index.js]: Store the lengths in constants instead of accessing them at each iteration.

File: README.md [hunk 28]
```diff
@@ -13,13 +13,15 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-nagios):
...
+the web interface, see [the plugin documentation](https://xen-orchestra.com/docs/plugins.html).
 
 ## Development
 
 ### `Xo#sendPassiveCheck(status, message) `
 
-This xo method is called to send a passive check to nagios and change the status of a service.
+This xo method is called to send a passive check to nagios and change the status of a service. It has two parameters:
+- status: it's the service status in Nagios (0: OK | 1: WARNING | 2: CRITICAL).
+- message: it's the status information in Nagios.
```
Comment (julien-f) [inline README.md:25]: Update the documentation to reflect the named params.

File: src/index.js [hunk 29]
```diff
@@ -71,65 +93,53 @@ class XoServerNagios {
     this._unset()
   }
 
-  test ({status}) {
-    return this._sendPassiveCheck(status, 'The server-nagios plugin for Xen Orchestra server seems to be working fine, nicely done :)')
+  test () {
+    return this._sendPassiveCheck({
+      status: 0,
```
Comment (julien-f) [inline src/index.js]: Export and use named constants:

```js
export const OK = 0
export const WARNING = 1
export const CRITICAL = 2
```

File: src/index.js [hunk 30]
```diff
@@ -71,65 +93,53 @@ class XoServerNagios {
...
+      message: 'The server-nagios plugin for Xen Orchestra server seems to be working fine, nicely done :)'
+    })
   }
 
-  _sendPassiveCheck (status, message) {
+  _sendPassiveCheck ({
+    status,
+    message
```
Comment (julien-f) [inline src/index.js]: Sort them.

File: src/index.js [hunk 31]
```diff
@@ -71,65 +93,53 @@ class XoServerNagios {
...
-    paquet.write(this._conf.service, 78, 206, encoding)
-    paquet.write(message, 206, size, encoding)
-    paquet.writeUInt32BE(crc32.unsigned(paquet), 4)
-    return paquet
-  }
+      client.on('data', data => {
+        const encoding = 'binary'
+        const timestamp = data.readInt32BE(128)
+        const packet = nscaPacketBuilder(this._conf.host, this._conf.service, SIZE, VERSION, timestamp, iv, status, message.toString().replace(/\s/g, ' '), encoding)
+        const iv = Buffer.from(data.toString(encoding, 0, 128), 'ascii') // initialization vector
```
Comment (julien-f) [inline src/index.js]: Avoid converting buffers back and forth.

If you just want a slice: use [`Buffer#slice()`](https://devdocs.io/node~4_lts/buffer#buffer_buf_slice_start_end).

File: src/index.js [hunk 32]
```diff
@@ -71,65 +93,53 @@ class XoServerNagios {
...
+        const key = Buffer.from(this._conf.key, 'ascii')
```
Comment (julien-f) [inline src/index.js]: This could be done once and for all in `configure()`.

File: src/index.js [hunk 33]
```diff
@@ -71,65 +93,53 @@ class XoServerNagios {
...
+          xor(
+            xor(
+              packet,
+              iv
+            ),
+            key
+          ),
+          encoding
+        )
+        client.write(xorPacketBuffer, (a) => {
```
Comment (julien-f) [inline src/index.js]: Do not use `()` around single param.

File: src/index.js [hunk 34]
```diff
@@ -71,65 +93,53 @@ class XoServerNagios {
...
+        })
+      })
 
-  _stringToAsciiArray (str) {
-    const map = Array.prototype.map
-    const chars = map.call(str, x => {
-      return x.charCodeAt(0)
+      client.on('error', (err) => {
+        reject(err)
+      })
```
Comment (julien-f) [inline src/index.js]: You can simplify: `client.on('error', reject)`.

File: README.md [hunk 35]
```diff
@@ -13,13 +13,15 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-nagios):
...
 Like all other xo-server plugins, it can be configured directly via
-the web interface, see [the plugin documentation](https://xen-orchestra.com/docs/plugins.html). You can also test if the configuration plugin works.
+the web interface, see [the plugin documentation](https://xen-orchestra.com/docs/plugins.html).
 
 ## Development
 
 ### `Xo#sendPassiveCheck(status, message) `
 
-This xo method is called to send a passive check to nagios and change the status of a service.
+This xo method is called to send a passive check to nagios and change the status of a service. It has two parameters:
```
Comment (julien-f) [inline README.md]: Add a black line between both sentences.

**Review by julien-f — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 23 above)
Comment (badrAZ) [reply src/index.js]: Nagios ignores all packets size under 720, and when i put a value over 720 i get an invalid CRC 32.

File: src/index.js (hunk 26 above)
Comment (badrAZ) [reply src/index.js]: It's an error, i will fix it.

File: src/index.js (hunk 23 above)
Comment (julien-f) [reply src/index.js]: Then the packet size should be 720 but defined directly in this function, no need to pass it as an argument.

Comment (badrAZ) [reply src/index.js]: ok, i will fix it.

File: src/index.js [hunk 36]
```diff
@@ -46,29 +56,34 @@ function nscaPacketBuilder (host, service, size, version, timestamp, iv, status,
...
+  packet.write(message, 206, SIZE, encoding)
   packet.writeUInt32BE(crc32.unsigned(packet), 4)
-  packet = Buffer.from(packet.toString(encoding), 'ascii')
   return packet
 }
 
-function xor (long, short) {
-  const result = []
+function xor (data, mask) {
+  const result = new Buffer(data.length)
```
Comment (julien-f) [inline src/index.js]: Use `dataSize` here.

File: src/index.js [hunk 37]
```diff
@@ -112,11 +129,17 @@ class XoServerNagios {
       })
 
       client.on('data', data => {
-        const encoding = 'binary'
         const timestamp = data.readInt32BE(128)
-        const packet = nscaPacketBuilder(this._conf.host, this._conf.service, SIZE, VERSION, timestamp, iv, status, message.toString().replace(/\s/g, ' '), encoding)
-        const iv = Buffer.from(data.toString(encoding, 0, 128), 'ascii') // initialization vector
-        const key = Buffer.from(this._conf.key, 'ascii')
+        const iv = new Buffer(data.slice(0, 128), ENCODING) // initialization vector
```
Comment (julien-f) [inline src/index.js]: This does not make sense:

- why copying the buffer returned by `slice()`?
- encoding should only be specified when creating a buffer from a string.

File: src/index.js [hunk 38]
```diff
@@ -35,9 +35,19 @@ export const configurationSchema = {
 
 // ===================================================================
 
-function nscaPacketBuilder (host, service, size, version, timestamp, iv, status, message, encoding) {
+function nscaPacketBuilder ({
+  encoding,
```
Comment (julien-f) [inline src/index.js]: How is the encoding a parameter for the function?

What values can be used?

File: src/index.js [hunk 39]
```diff
@@ -112,11 +129,17 @@ class XoServerNagios {
...
         const timestamp = data.readInt32BE(128)
-        const packet = nscaPacketBuilder(this._conf.host, this._conf.service, SIZE, VERSION, timestamp, iv, status, message.toString().replace(/\s/g, ' '), encoding)
-        const iv = Buffer.from(data.toString(encoding, 0, 128), 'ascii') // initialization vector
-        const key = Buffer.from(this._conf.key, 'ascii')
+        const iv = new Buffer(data.slice(0, 128), ENCODING) // initialization vector
+        const packet = nscaPacketBuilder({
+          ...this._conf,
+          encoding: ENCODING,
+          iv,
+          message: message.toString().replace(/\s/g, ' '),
```
Comment (julien-f) [inline src/index.js]: Why `toString()`? Isn't `message` already a string?

Comment (julien-f) [inline src/index.js]: Why replacing whitespaces by a single space?

File: src/index.js [hunk 40]
```diff
@@ -35,9 +35,19 @@ export const configurationSchema = {
...
-function nscaPacketBuilder (host, service, size, version, timestamp, iv, status, message, encoding) {
+function nscaPacketBuilder ({
+  encoding,
+  host,
+  iv,
+  message,
+  service,
+  status,
+  timestamp,
+  version
```
Comment (julien-f) [inline src/index.js]: Should not be a param, simply use `VERSION`.

File: src/index.js [hunk 41]
```diff
@@ -1,26 +1,166 @@
...
+          iv,
+          message: message.toString().replace(/\s/g, ' '),
+          status,
+          timestamp,
+          version: VERSION
+        })
+
+        // 1) Using xor between the NSCA packet and the initialization vector
+        // 2) Using xor between the result of the first operation and the encryption key
+        const xorPacketBuffer = new Buffer(
```
Comment (julien-f) [inline src/index.js]: Why are you copying the buffer returned by `xor()`?

File: src/index.js [hunk 42]
```diff
@@ -1,26 +1,166 @@
...
+          xor(
+            xor(
+              packet,
+              iv
+            ),
+            this._key
+          ),
+          ENCODING
```
Comment (julien-f) [inline src/index.js]: Does not make sense.

**Review by julien-f — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 39 above)
Comment (badrAZ) [reply src/index.js]: It's an error, i will fix it.

Comment (badrAZ) [reply src/index.js]: Because Nagios doesn't appreciate a multiple lines in the message.

File: src/index.js (hunk 41 above)
Comment (badrAZ) [reply src/index.js]: It's an error i will fix it.

File: src/index.js (hunk 38 above)
Comment (badrAZ) [reply src/index.js]: It's an error. I will use the constant instead.

File: src/index.js (hunk 39 above)
Comment (badrAZ) [reply src/index.js]: After thought, i think it's not necessary  because the message is short.

Comment (julien-f) [reply src/index.js]: If the message should not contain a line break, you should test for it and throw an error.

File: src/index.js [hunk 43]
```diff
@@ -121,6 +119,10 @@ class XoServerNagios {
     message,
     status
   }) {
+    if (/\r|\n/.exec(message)) {
```
Comment (julien-f) [inline src/index.js]: Use `RegExp#exec()` to capture substrings or `RegExp#test()` if you just want to validate the string.

File: src/index.js [hunk 44]
```diff
@@ -121,6 +119,10 @@ class XoServerNagios {
...
+      throw new Error('The message must not contain a line break.')
```
Comment (julien-f) [inline src/index.js]: Error messages are not sentences: they should neither start with an upper-case letter nor end with a period.

**Review by julien-f — CHANGES_REQUESTED** [review]

File: src/index.js [hunk 45]
```diff
@@ -1,26 +1,164 @@
...
+  packet.fill(0)
+  packet.writeInt16BE(VERSION, 0)
+  packet.fill('h', 2, 3)
+  packet.writeUInt32BE(0, 4) // initial 0 for CRC32 value
+  packet.writeUInt32BE(timestamp, 8)
+  packet.writeInt16BE(status, 12)
+  packet.write(host, 14, 77, ENCODING)
+  packet.write(service, 78, 206, ENCODING)
+  packet.write(message, 206, SIZE, ENCODING)
+  packet.writeUInt32BE(crc32.unsigned(packet), 4)
```
Comment (julien-f) [inline src/index.js:58]: No need to convert to and from an uint32: `crc32()` directly return a buffer.

File: src/index.js [hunk 46]
```diff
@@ -1,26 +1,164 @@
...
+
+  _sendPassiveCheck ({
+    message,
+    status
+  }) {
+    if (/\r|\n/.test(message)) {
+      throw new Error('the message must not contain a line break')
+    }
+
+    const client = new net.Socket()
```
Comment (julien-f) [inline src/index.js]: Move all the code inside the promise to avoid throwing synchronously.

**Review by julien-f — CHANGES_REQUESTED** [review]

**Review by julien-f — APPROVED** [review]

**Comment by julien-f** [comment]
@olivierlambert The code is OK for me. What do you think of the functional side?

**Comment by olivierlambert** [comment]
I'll test on my side. Can you assign me for a review?

**Review by olivierlambert — APPROVED** [review]
LGTM

**Pull request merged**